Fix switcher_kis services access

yinonco opened the pull request:
…service

this permissions validation is preventing the option of turning on the device from a script or an automation

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I deleted the permissions validation from one of the services this integration (switcher_kis) exposes.
The service is `switcher_kis.turn_on_with_timer` and the reason is that this permissions validation is not allowing to turn on the device from an automation or a script.
This error was reported here: https://community.home-assistant.io/t/error-running-service-from-automation/218722
and I have encountered this as well.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: No issue was opened for this but people reported an this error before here:
https://community.home-assistant.io/t/error-running-service-from-automation/218722
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

yinonco committed:
deleted _validate_edit_permission call from async_turn_on_with_timer_service

this permissions validation is preventing the option of turning on the device from a script or automation

homeassistant commented:
Hi @yinonco, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#39693).

Once you do that we will be able to review and accept this pull request.

Thanks!

yinonco committed:
deleted _validate_edit_permission and its usages

to allow all async_set_auto_off_service service to run from script/automation as well

yinonco commented:
deleted all permissions validations to allow `set_auto_off_service` to run from acript/automation as well

yinonco committed:
deleted unused imports

yinonco committed:
deleted irrelevant tests

yinonco committed:
deleted unused imports

TomerFi commented:
As far as I can see, the permission module is still marked experimental,
as it was at the time of writing the integration:
https://developers.home-assistant.io/docs/auth_permissions

I'm not actively involved or following Home Assistant's development,
So I have no idea what's the status of the permission module today.

If there are any plans of bumping it out of experimental status,
Then maybe removing its usage completely from this integration, isn't the right way to go.

yinonco commented:
Agree, but its current implementation isn't correct anyway because a. there is no way to use it with automations and scripts (which also means that I cannot bind an input slider to the boiler timer - so it makes it kind of useless in most cases) and b. in the current implementation you can simply give it any entity which starts with the expected prefix and which you have permissions on and it will turn on the boiler even if you don't have permissions to the boiler entity

TomerFi commented:
As I said,
If there are any plans for enforcing the `permission module` any time soon.
Then I'm not sure the correct way of handling the permission issue with `switcher integration`,
is by removing the implementation of the `permission module` altogether.

- Maybe this issue has already been resolved and can be fixed with proper implementation in `switcher integration`,
  After all, the implementation was done based on the documentation at the time,
  which probably differs a lot since then. How do other integrations implement this module?

- Maybe this is issue, effects other integrations too and should be resolved with the `permission module`,

- Maybe there are no plans of promoting the `permission module`, in this case, this PR seems legit to me.

As stated in my previous comment,
I'm not actively involved or following `Home Assistant`'s development,
I was asked to review this code by the `code owners bot`.

As I'm not sure what's the current status of, or the "correct way of implementation" is, for the `permission module`,
It wouldn't be right for me to review a PR which removes the use of it,
without doing some proper homework first.

I'm sure someone else with more knowledge then me on this topic can review this PR.
I'll keep on following, in case there will be something else I can help with besides my opinion written above.

yinonco commented:
I checked other components and I didnt see something similar.
The problem is not with the permissions module but with the implementation here
This is part of the code which I deleted:
`    if user is None:`
`        raise UnknownUser(context=context, entity_id=entity_id, permission=POLICY_EDIT)
`
when you call a service from an automation you do not have a user - so you will always get `UnknownUser`

TomerFi commented:
> This is part of the code which I deleted:
if user is None:
raise UnknownUser(context=context, entity_id=entity_id, permission=POLICY_EDIT)
when you call a service from an automation you do not have a user - so you will always get UnknownUser

You've deleted the whole `permission module` implementation, and not just that part.

> **plus** as I said before the implementation here is wrong for another reason - you get the `entity id` from outside and use it only to check user permissions - so if a user don't have permissions to the switcher entity but does have to another entity with the same prefix he could send the entity_id which he has permissions on and the code would validate the permissions successfully and will call the service on the **actual** boiler entity and not the entity he sent!!!

As I said,
The implementation of the `permission module` in `switcher integration` was done based on the documentation at the time,
which was more than a year ago. A year for an experimental feature means that probably a lot has changed since then.
**Please feel free to contribute and implement it the right way!!!**

By the way, the `swithcer integration` was introduced with version `0.93` on May 16th, 2019, [here](https://www.home-assistant.io/blog/2019/05/16/release-93/).
Based on the information you provided, the issue was first reported on August 15th, 2020, as far as I can see [here](https://community.home-assistant.io/t/error-running-service-from-automation/218722/9), by one user.

Your solution is to **remove the usage of the `permission module` altogether** because the "implementation here is wrong" and "it makes the service almost useless", yet you didn't try to implement the right way and making it useful, and you didn't attach any link to a topic that might imply that the `permission module` is not going to get promoted out of experimental mode.

Maybe you're right, as I said a couple of times in this conversation, this was per-documentation at the time.
Please note that the `permission module` is still in an experimental module.

As I'm not actively following `Home Assistant` please hold for someone with more knowledge of the `permission module` than me to assist you.
And please stop sending comments that end with `!!!` my way,
you're just making it harder for me to want to help.

yinonco commented:
I have read the permissions documentation now.
I didnt find anything helpful for this problem.
Furthermore, I searched all other core components for similar permissions implementation and found one
so another solution I can suggest would be to treat user None as positive permissions (saw this in other components `amcrest`) for example.

`def have_permission(user, entity_id):`
`        return not user or user.permissions.check_entity(entity_id, POLICY_CONTROL)`

yinonco commented:
and BTW I think it should be POLICY_CONTROL and not POLICY_EDIT so I can change that as well

MartinHjelmare commented on homeassistant/components/switcher_kis/__init__.py:
```diff
@@ -78,21 +74,6 @@
-) -> None:
-    """Use for validating user control permissions."""
-    splited = split_entity_id(entity_id)
-    if splited[0] != SWITCH_DOMAIN or not splited[1].startswith(DOMAIN):
```
Although looking at the forum post log, the problem seems to be that the user is unknown.

yinonco replied:
First of all, thanks a lot for your time.

Another problem is that the integration is implemented in such way that it supports only one device
and if you look in the actual `turn_on_with_timer` service's code you see that there is no reference to the `entity_id` while calling to the device's API, actually this is the same for both services, you can see it in `__init__.py` under `async_set_auto_off_service` and `async_turn_on_with_timer_service`:

`async with SwitcherV2Api( hass.loop, device_data.ip_addr, phone_id, device_id, device_password ) as swapi:`
`await swapi.control_device(COMMAND_ON, service.data[CONF_TIMER_MINUTES])`

So, even if I will use entity service, as long as the integration supports only one device and the entity_id is not taken into consideration while calling the device API - In case I will call the service with my `switch.room_light` entity_id it will turn on the boiler, regardless what entity I sent to the service.

So I need some way to enable the service only for `switcher` devices (same as we cannot call `light.turn_on` on a `switch` entity)
And as long as there is no support for multiple devices and the service implementation will turn on the same device regardless what entity_id was sent - I think we can remove the `entity_id` var from the service.

Feel free to correct me if I wrong :)

yinonco replied:
Ok I will give it a try

royi1000 commented:
> Hi @yinonco,
> 
> It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant/core#39693).
> 
> Once you do that we will be able to review and accept this pull request.
> 
> Thanks!

CLA signed, please merge

leranp commented:
I have tested the commit and now everything works, please approve

frenck commented:
This PR has been superseded by #45204
